lte/gateway/c/session_manager/SetMessageManagerHandler.cpp

```diff
@@ -504,43 +504,39 @@ void SetMessageManagerHandler::service_handle_request_on_paging(
     std::function<void(Status, SmContextVoid)> response_callback) {
   // extract IMSI value from proto
   auto imsi = notif.common_context().sid().id();
+  uint32_t pdu_id = notif.rat_specific_notification().pdu_session_id();
+  SessionSearchCriteria criteria(imsi, IMSI_AND_PDUID, pdu_id);
   auto session_map = session_store_.read_sessions({imsi});
-  int count = 0;
-  auto session_update = SessionStore::get_default_session_update(session_map);
-  for (auto& session : session_map[imsi]) {
-    if (session->get_state() == INACTIVE) {
-      auto session_id = session->get_session_id();
-      SessionStateUpdateCriteria& session_uc = session_update[imsi][session_id];
-      m5g_enforcer_->m5g_move_to_active_state(session, notif, &session_uc);
-      bool update_success = session_store_.update_sessions(session_update);
-      if (!update_success) {
-        MLOG(MINFO) << "Operation aborted in  middle"
-                    << " session id: " << session->get_session_id() << imsi
-                    << " of imsi";
-        continue;
-      }
-      MLOG(MDEBUG) << "Successfully updated SessionStore "
-                   << " session id: " << session->get_session_id() << imsi
-                   << " of imsi";
-      count++;
-    }
-  }
-  if (!count) {
+  auto session_it = session_store_.find_session(session_map, criteria);
+  if (!session_it) {
     MLOG(MINFO) << " No Valid session found for IMSI: " << imsi;
     Status status(grpc::NOT_FOUND, "Sesion Not found");
     response_callback(status, SmContextVoid());
     return;
   }
-  bool update_success = session_store_.update_sessions(session_update);
-  if (!update_success) {
-    Status status(grpc::ABORTED, "update operation aborted in  middle");
-    response_callback(status, SmContextVoid());
-    return;
+  auto& session = **session_it;
+  auto session_update = SessionStore::get_default_session_update(session_map);
+  if (session->get_state() == INACTIVE) {
+    auto session_id = session->get_session_id();
+    SessionStateUpdateCriteria& session_uc = session_update[imsi][session_id];
+    m5g_enforcer_->m5g_move_to_active_state(session, notif, &session_uc);
+    bool update_success = session_store_.update_sessions(session_update);
+    if (!update_success) {
+      MLOG(MINFO) << "Operation aborted in  middle"
+                  << " session id: " << session->get_session_id() << imsi
+                  << " of imsi";
+      Status status(grpc::ABORTED, "update operation aborted in  middle");
+      response_callback(status, SmContextVoid());
+      return;
+    } else {
+      MLOG(MDEBUG) << "Successfully updated SessionStore "
+                   << " session id: " << session->get_session_id() << imsi
+                   << " of imsi";
+    }

```

Comment:
add warning message in case session is active and paging was received.

Reply:
In `UpfMsgManageHandler.cpp` if UPF receives paging request under `get_session_from_imsi` method we will check the session state is `INACTIVE` or not, if it is `INACTIVE` then only we will forward this notification to AMF via `conv_enforcer_->handle_state_update_to_amf` method, else we will trigger below message and does nothing for this session.
```Dec 9 02:21:56 TVM3-SA-dev sessiond[3166004]: I1209 02:21:55.974514 3166004 UpfMsgManageHandler.cpp:239] Can not Trigger Paging notification to AMF, as session is not an INACTIVE state.```
So, If session state is anything other than `INACTIVE` paging will not trigger for that session and we will receive above message.

In the file `SetMessageManagerHandler.cpp` the above code from method `service_handle_request_on_paging.` This function called when specific IMSI service request is received from AMF. By this time session state should be INACTIVE only. here no other states will not come. I believe adding code check for any other state would be Unreachable code here. Please let me know your thoughts here.

Reply:
Does it make sense UPF receives a paging request when the UE is not IDLE 
Shouldnt this fact be considered as something potentially bad (for example if we keep receiving pagings and we do not show them on the log we may be hiding an issue from the core)?

Reply:
UPF only trigger paging when paging flow is  present in table0 entries of OVS. So only SMF receive the paging notification when UE is in IDLE state [INACTIVE state]